# Toronto Neighborhood Clustering Analysis

In [2]:
import numpy as np 

import pandas as pd 
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

from bs4 import BeautifulSoup

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - folium=0.5.0


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    branca-0.3.1               |             py_0          25 KB  conda-forge
    vincent-0.4.4              |             py_1          28 KB  conda-forge
    openssl-1.1.1c             |       h516909a_0         2.1 MB  conda-forge
    certifi-2019.6.16          |           py36_1         149 KB  conda-forge
    folium-0.5.0               |             py_0          45 KB  conda-forge
    altair-3.1.0               |           py36_0         724 KB  conda-forge
    ca-certificates-2019.6.16  |       hecc5488_0         145 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         3.2 MB

The following NEW packages will be 

Using BeautifulSoup we get the [table of Toronto neighborhods by area code](https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M) from Wikipedia. We will build the pandas Data Frame 'df" which countains the Postcode, Borough, and Neighborhoods from this table.

In [5]:
website_url = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text
soup = BeautifulSoup(website_url,'lxml')
table = soup.find('table',{'class':'wikitable sortable'})
ths = table.find_all('th')

In [6]:
headings = [th.text.strip() for th in ths]
headings[2] = 'Neighborhood'

Table_Data = []
for tr in table.find_all('tr'):
    tds = tr.find_all('td')
    if not tds:
        continue
    Postcode, Borough, Neighborhood = [td.text.strip() for td in tds[:3]]
    Table_Data.append([Postcode, Borough, Neighborhood])
    
Table_Data_Grouped = []
for row in Table_Data:
    entry = []

df = pd.DataFrame(Table_Data)
df.columns = headings 

df = df[df.Borough != 'Not assigned'] # This removes all entries which have no Borough name

df.loc[df['Neighborhood'] == 'Not assigned','Neighborhood'] = df['Borough'] # This replaces Neighborhood name with Borough name if Neighborhood name is not assigned

neib = lambda a: ", ".join(a)
df = df.groupby(by='Postcode').agg({'Borough': 'max','Neighborhood': neib}).reset_index() # This groups neighborhoods with the same postcode

df.head()

Postcode      Borough                            Neighborhood
0      M1B  Scarborough                          Rouge, Malvern
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union
2      M1E  Scarborough       Guildwood, Morningside, West Hill
3      M1G  Scarborough                                  Woburn
4      M1H  Scarborough                               Cedarbrae

In [7]:
df.shape

(103, 3)

Using geospatial data we will append two additional columns for 'Latitude' and 'Longitude' of each borough to the dataframe, and rename the dataframe 'df2.'

In [8]:
url = 'https://cocl.us/Geospatial_data'
geo=pd.read_csv(url)
df2 = pd.concat([df, geo['Latitude'], geo['Longitude']], axis=1, sort=False)
df2.head()

Postcode      Borough                            Neighborhood   Latitude  \
0      M1B  Scarborough                          Rouge, Malvern  43.806686   
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2      M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3      M1G  Scarborough                                  Woburn  43.770992   
4      M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

Using folium, we will construct a map of Toronto.  We will superimpose borough markers on top of the map using data from 'df2.' 

In [9]:
address = 'Toronto, ON'

geolocator = Nominatim(user_agent="Toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough in zip(df2['Latitude'], df2['Longitude'], df2['Borough']):
    label = '{}'.format(borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

We will query the Foursquare API for venue data corresponding to the Toronto boroughs in 'df2.'  The function below runs through the boroughs in 'df2' and queries the API, appending the query results to venues_list.  These will be aggregated into the dataframe toronto_venues.

In [10]:
CLIENT_ID = 'WQZ4C3RX1UX3YLTANCQFQ4XMOD0W5LSMRHJSTCGXJQ0VEX1L' # your Foursquare ID
CLIENT_SECRET = 'BX10JIX4Q0R3JACDMIQXFHINLCOXPUDWNWERJOCVFCGZRWX5' # your Foursquare Secret
VERSION = '20190719' # Foursquare API version

def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Borough', 
                  'Latitude', 
                  'Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

LIMIT = 100

radius = 500

toronto_venues = getNearbyVenues(names=df2['Borough'],
                                   latitudes=df2['Latitude'],
                                   longitudes=df2['Longitude']
                                  )


toronto_venues.head()

Borough   Latitude  Longitude                            Venue  \
0  Scarborough  43.806686 -79.194353                          Wendy's   
1  Scarborough  43.806686 -79.194353            Interprovincial Group   
2  Scarborough  43.784535 -79.160497            Royal Canadian Legion   
3  Scarborough  43.763573 -79.188711  Swiss Chalet Rotisserie & Grill   
4  Scarborough  43.763573 -79.188711                G & G Electronics   

   Venue Latitude  Venue Longitude        Venue Category  
0       43.807448       -79.199056  Fast Food Restaurant  
1       43.805630       -79.200378            Print Shop  
2       43.782533       -79.163085                   Bar  
3       43.767697       -79.189914           Pizza Place  
4       43.765309       -79.191537     Electronics Store

In [11]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 269 uniques categories.


The 269 unique venue categries will be one hot encoded for the boroughs, and a new dataframe will be created with the venue information.  We will name this dataframe 'toronto_onehot.'  Then we will group by neighborhood into the dataframe 'toronto_grouped.'

In [14]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Borough'] = toronto_venues['Borough'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_grouped = toronto_onehot.groupby('Borough').mean().reset_index()
toronto_grouped.head()

Borough  Accessories Store  Afghan Restaurant   Airport  \
0   Central Toronto                0.0           0.000000  0.000000   
1  Downtown Toronto                0.0           0.000772  0.000772   
2      East Toronto                0.0           0.000000  0.000000   
3         East York                0.0           0.000000  0.000000   
4         Etobicoke                0.0           0.000000  0.000000   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0            0.000000      0.000000        0.000000         0.000000   
1            0.000772      0.000772        0.001543         0.002315   
2            0.000000      0.000000        0.000000         0.000000   
3            0.000000      0.000000        0.000000         0.000000   
4            0.000000      0.000000        0.000000         0.000000   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0          0.000000             0.018182      0.000000  0.000000     0.000000   
1          0.001543             0.014660      0.001543  0.003858     0.008488   
2          0.000000             0.024590      0.000000  0.000000     0.000000   
3          0.000000             0.000000      0.000000  0.000000     0.000000   
4          0.000000             0.014286      0.000000  0.000000     0.000000   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0    0.000000             0.000000          0.000000            0.000000   
1    0.001543             0.002315          0.006173            0.000772   
2    0.000000             0.000000          0.000000            0.000000   
3    0.000000             0.000000          0.013333            0.026667   
4    0.000000             0.000000          0.000000            0.014286   

   Auto Garage  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop    Bakery  \
0          0.0       0.000000   0.009091    0.000000    0.009091  0.000000   
1          0.0       0.000000   0.000772    0.000772    0.003086  0.024691   
2          0.0       0.008197   0.000000    0.000000    0.000000  0.024590   
3          0.0       0.000000   0.000000    0.000000    0.013333  0.000000   
4          0.0       0.000000   0.000000    0.000000    0.000000  0.028571   

       Bank       Bar  Baseball Field  Baseball Stadium  Basketball Court  \
0  0.000000  0.000000        0.000000          0.000000               0.0   
1  0.003086  0.019290        0.000000          0.001543               0.0   
2  0.008197  0.008197        0.000000          0.000000               0.0   
3  0.040000  0.000000        0.000000          0.000000               0.0   
4  0.014286  0.000000        0.014286          0.000000               0.0   

   Basketball Stadium     Beach  Beer Bar  Beer Store  Belgian Restaurant  \
0            0.000000  0.000000  0.000000    0.000000            0.000000   
1            0.003086  0.000772  0.013889    0.002315            0.002315   
2            0.000000  0.000000  0.000000    0.000000            0.000000   
3            0.000000  0.000000  0.000000    0.026667            0.000000   
4            0.000000  0.000000  0.000000    0.028571            0.000000   

   Bike Shop    Bistro  Board Shop  Boat or Ferry  Bookstore  Boutique  \
0   0.000000  0.000000    0.000000       0.000000   0.000000  0.000000   
1   0.000000  0.003086    0.000000       0.000772   0.009259  0.000772   
2   0.000000  0.000000    0.008197       0.000000   0.016393  0.000000   
3   0.013333  0.000000    0.000000       0.000000   0.000000  0.000000   
4   0.000000  0.000000    0.000000       0.000000   0.000000  0.000000   

   Brazilian Restaurant  Breakfast Spot   Brewery  Bridal Shop  \
0              0.000000        0.009091  0.009091          0.0   
1              0.001543        0.011574  0.003858          0.0   
2              0.000000        0.000000  0.032787          0.0   
3              0.000000        0.026667  0.013333          0.0   
4              0.000000        0.00000

For illustration purposes, we will show the top five venue categories in each borough below. 

In [15]:
num_top_venues = 5

for hood in toronto_grouped['Borough']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Borough'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Central Toronto----
            venue  freq
0     Coffee Shop  0.08
1     Pizza Place  0.06
2  Sandwich Place  0.06
3            Park  0.06
4            Café  0.05


----Downtown Toronto----
         venue  freq
0  Coffee Shop  0.09
1         Café  0.05
2   Restaurant  0.03
3        Hotel  0.03
4         Park  0.02


----East Toronto----
                venue  freq
0         Coffee Shop  0.07
1    Greek Restaurant  0.07
2  Italian Restaurant  0.05
3                Café  0.03
4             Brewery  0.03


----East York----
         venue  freq
0  Coffee Shop  0.08
1         Park  0.05
2  Pizza Place  0.05
3         Bank  0.04
4     Pharmacy  0.04


----Etobicoke----
                  venue  freq
0           Pizza Place  0.11
1        Sandwich Place  0.07
2              Pharmacy  0.06
3           Coffee Shop  0.06
4  Fast Food Restaurant  0.04


----Mississauga----
                       venue  freq
0                      Hotel  0.18
1                Coffee Shop  0.18
2        Fried 

In [16]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [17]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Borough']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
boroughs_venues_sorted = pd.DataFrame(columns=columns)
boroughs_venues_sorted['Borough'] = toronto_grouped['Borough']

for ind in np.arange(toronto_grouped.shape[0]):
    boroughs_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

boroughs_venues_sorted.head()

Borough 1st Most Common Venue 2nd Most Common Venue  \
0   Central Toronto           Coffee Shop           Pizza Place   
1  Downtown Toronto           Coffee Shop                  Café   
2      East Toronto      Greek Restaurant           Coffee Shop   
3         East York           Coffee Shop           Pizza Place   
4         Etobicoke           Pizza Place        Sandwich Place   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0        Sandwich Place                  Park                  Café   
1            Restaurant                 Hotel                Bakery   
2    Italian Restaurant               Brewery        Ice Cream Shop   
3                  Park              Pharmacy                  Bank   
4           Coffee Shop              Pharmacy                  Café   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0      Sushi Restaurant          Dessert Shop                   Pub   
1    Italian Restaurant                   Bar   Japanese Restaurant   
2                  Café           Yoga Studio                   Pub   
3          Burger Joint   Sporting Goods Shop                   Gym   
4  Fast Food Restaurant         Grocery Store                   Gym   

  9th Most Common Venue 10th Most Common Venue  
0        Clothing Store                    Gym  
1                  Park             Steakhouse  
2           Pizza Place                 Bakery  
3         Grocery Store      Indian Restaurant  
4                  Park           Liquor Store

Now we will segement and cluster the neighborhood into 5 clusters determined by the venues appearing commonly. This will be done using k-means clustering.

In [18]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Borough', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 0, 0, 1, 1, 3, 0, 4, 0, 0], dtype=int32)

In [19]:
toronto_grouped_clustering.head()

Accessories Store  Afghan Restaurant   Airport  Airport Food Court  \
0                0.0           0.000000  0.000000            0.000000   
1                0.0           0.000772  0.000772            0.000772   
2                0.0           0.000000  0.000000            0.000000   
3                0.0           0.000000  0.000000            0.000000   
4                0.0           0.000000  0.000000            0.000000   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0      0.000000        0.000000         0.000000          0.000000   
1      0.000772        0.001543         0.002315          0.001543   
2      0.000000        0.000000         0.000000          0.000000   
3      0.000000        0.000000         0.000000          0.000000   
4      0.000000        0.000000         0.000000          0.000000   

   American Restaurant  Antique Shop  Aquarium  Art Gallery  Art Museum  \
0             0.018182      0.000000  0.000000     0.000000    0.000000   
1             0.014660      0.001543  0.003858     0.008488    0.001543   
2             0.024590      0.000000  0.000000     0.000000    0.000000   
3             0.000000      0.000000  0.000000     0.000000    0.000000   
4             0.014286      0.000000  0.000000     0.000000    0.000000   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Garage  \
0             0.000000          0.000000            0.000000          0.0   
1             0.002315          0.006173            0.000772          0.0   
2             0.000000          0.000000            0.000000          0.0   
3             0.000000          0.013333            0.026667          0.0   
4             0.000000          0.000000            0.014286          0.0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop    Bakery      Bank  \
0       0.000000   0.009091    0.000000    0.009091  0.000000  0.000000   
1       0.000000   0.000772    0.000772    0.003086  0.024691  0.003086   
2       0.008197   0.000000    0.000000    0.000000  0.024590  0.008197   
3       0.000000   0.000000    0.000000    0.013333  0.000000  0.040000   
4       0.000000   0.000000    0.000000    0.000000  0.028571  0.014286   

        Bar  Baseball Field  Baseball Stadium  Basketball Court  \
0  0.000000        0.000000          0.000000               0.0   
1  0.019290        0.000000          0.001543               0.0   
2  0.008197        0.000000          0.000000               0.0   
3  0.000000        0.000000          0.000000               0.0   
4  0.000000        0.014286          0.000000               0.0   

   Basketball Stadium     Beach  Beer Bar  Beer Store  Belgian Restaurant  \
0            0.000000  0.000000  0.000000    0.000000            0.000000   
1            0.003086  0.000772  0.013889    0.002315            0.002315   
2            0.000000  0.000000  0.000000    0.000000            0.000000   
3            0.000000  0.000000  0.000000    0.026667            0.000000   
4            0.000000  0.000000  0.000000    0.028571            0.000000   

   Bike Shop    Bistro  Board Shop  Boat or Ferry  Bookstore  Boutique  \
0   0.000000  0.000000    0.000000       0.000000   0.000000  0.000000   
1   0.000000  0.003086    0.000000       0.000772   0.009259  0.000772   
2   0.000000  0.000000    0.008197       0.000000   0.016393  0.000000   
3   0.013333  0.000000    0.000000       0.000000   0.000000  0.000000   
4   0.000000  0.000000    0.000000       0.000000   0.000000  0.000000   

   Brazilian Restaurant  Breakfast Spot   Brewery  Bridal Shop  \
0              0.000000        0.009091  0.009091          0.0   
1              0.001543        0.011574  0.003858          0.0   
2              0.000000        0.000000  0.032787          0.0   
3              0.000000        0.026667  0.013333          0.0   
4              0.000000        0.000000  0.000000          0.0   

   Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \

In [20]:
# add clustering labels
boroughs_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = df2

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(boroughs_venues_sorted.set_index('Borough'), on='Borough')

toronto_merged.tail() # check the last columns!

Postcode    Borough                                       Neighborhood  \
98       M9N       York                                             Weston   
99       M9P  Etobicoke                                          Westmount   
100      M9R  Etobicoke  Kingsview Village, Martin Grove Gardens, Richv...   
101      M9V  Etobicoke  Albion Gardens, Beaumond Heights, Humbergate, ...   
102      M9W  Etobicoke                                          Northwest   

      Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
98   43.706876 -79.518188               2                  Park   
99   43.696319 -79.532242               1           Pizza Place   
100  43.688905 -79.554724               1           Pizza Place   
101  43.739416 -79.588437               1           Pizza Place   
102  43.706748 -79.594054               1           Pizza Place   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
98      Convenience Store               Brewery  Fast Food Restaurant   
99         Sandwich Place           Coffee Shop              Pharmacy   
100        Sandwich Place           Coffee Shop              Pharmacy   
101        Sandwich Place           Coffee Shop              Pharmacy   
102        Sandwich Place           Coffee Shop              Pharmacy   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
98                  Field        Sandwich Place        Discount Store   
99                   Café  Fast Food Restaurant         Grocery Store   
100                  Café  Fast Food Restaurant         Grocery Store   
101                  Café  Fast Food Restaurant         Grocery Store   
102                  Café  Fast Food Restaurant         Grocery Store   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
98             Restaurant  Caribbean Restaurant            Coffee Shop  
99                    Gym                  Park           Liquor Store  
100                   Gym                  Park           Liquor Store  
101                   Gym                  Park           Liquor Store  
102                   Gym                  Park           Liquor Store

To visualize this clustering, we will redraw the map of Toronto and color the borough markers according to cluster. 

In [22]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Borough'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [ ]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

In [ ]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

In [ ]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

In [ ]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

In [ ]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]